server/src/main/java/io/crate/execution/engine/aggregation/impl/average/IntegralAverageState.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.engine.aggregation.impl.average;
+
+import javax.annotation.Nonnull;
+
+public class IntegralAverageState extends AverageState {
+
+    @Override
+    public void addNumber(Number number) {
+        this.sum += number.doubleValue();
+        this.count++;
```

Comment:
Since we split up the two states, I think we could also make `sum` a `long` to switch from floating point addition instructions to `ladd`

Reply:
it will be changed in `OverflowAwareMutableLong` in #11612 - neither double nor long are capable to handle case of 2 `MAX_LONG`-s as nominator will overflow/lose precision. `OverflowAwareMutableLong` would extend `Number` and `double sum`  would be changed to `Number sum`. If we don't want to change return type in 4.7 (at least `bigint` as `avg` belongs to type range and cannot exceed `bigint` for integrals or even numeric to align with PG) then it doesn't make sense to exchange one issue to another.

Reply:
I'm thinking we could handle this similarly to how we handled sum, where we have a dedicated NumericSumAggregation and where we require opt-in from the user via explicit cast to numeric to use it. This would avoid the change in behavior and we could optimize the "sum some int values" cases to use integer arithmetic instead of floating arithmetic.

Reply:
The thing is final average value is not supposed to overflow, it's intermediate nominator sum being the issue - wouldn't it be annoying to users to cast? I think we can go for explicit cast opt-in as a current fix (and add it to docs like we did for sum) but probably would need a follow up in 5.0 where we can afford having breaking change of return type.

If we go with explicit cast opt-in question what do we do in case of overflow - throw ArithmeticException or just warn user that they must cast otherwise they got incorrect result. I feel like throwing exception is more honest than providing wrong result but then again it's behavioural change.

Reply:
Not sure if the overflowing behavior of double is better than that of long and this PR already introduces the `IntegralAverageState`, but fine for me to address this in a follow up